DDD-8.md

```diff
@@ -126,3 +126,146 @@ So if the snapshot is reading a chunk from a table that is modified through a su
+
+This type of configuration is common for CDC scenario where users wants to reduce the load on the primary that is already under the write load.
+
+So you can directly connect to a secondary specifying its address and setting the read preference mode accordingly. 
+
+This configuration can lead to a possible weird issues in particular type of deployments. For example, if you have that nodes of the RS has two network address, internal and external, and you configure 
+the connection string to point to a secondary using its external address, when the client connects to the RS and retrieve its information, even the intern addresses of the nodes, it then can use the internal address causing connection failures. 
+To overcome to this, you can set the [directConnection](https://www.mongodb.com/docs/manual/reference/connection-string/#mongodb-urioption-urioption.directConnection) property.
+
+##### Sharded cluster
+
+A [MongoDB sharded cluster](https://www.mongodb.com/docs/manual/core/sharded-cluster-components/) (SC) consists of the following components:
+
+* shard: Each shard contains a subset of the sharded data. Each shard must be deployed as a replica set.
+
+* mongos: The mongos acts as a query router, providing an interface between client applications and the sharded cluster. mongos can support hedged reads to minimize latencies.
+
+* config servers: Config servers store metadata and configuration settings for the cluster. Config servers must be deployed as a replica set (CSRS).
+
+The standard way to connect to SC is using the following connection string
+
+```text
+mongodb://user:password@mongos0.example.com:27017,mongos1.example.com:27017,mongos2.example.com:27017/?authSource=admin
+```
+
+#### Oplog and Change stream
+
+The [oplog](https://www.mongodb.com/docs/manual/core/replica-set-oplog/) (operations log) is a special capped collection that keeps a rolling record of all operations that modify the data stored in your databases.
+If write operations do not modify any data or fail, they do not create oplog entries. 
+
+MongoDB applies database operations on the primary and then records the operations on the primary's oplog. The secondary members then copy and apply these operations in an asynchronous process. All replica set members contain a copy of the oplog, in the local.oplog.rs collection, which allows them to maintain the current state of the database.
+To facilitate replication, all replica set members send heartbeats (pings) to all other members. Any secondary member can import oplog entries from any other member.
+
+Change streams allow applications to access real-time data changes without the prior complexity and risk of manually tailing the oplog. Applications can use change streams to subscribe to all data changes on a single collection, a database, or an entire deployment, and immediately react to them.
+Change streams is used by MongoDB Debezium connector. 
```

Comment:
Note that mongos itself has no oplog, however it still can be used to open a change stream. Under the hood mongos will open change stream against each shard, synchronise the events and serve them to the client via the change stream opened against mongos.

The need for synchronisation can be a source of performance degradation if there is a so called "cold shard" (a shard with minimal write) as mongos need to check whether the shard received any new changes -- no change is noted on the oplog via periodic noop entry in the oplog. So potentially the synchronisation can be delayed until the next noop is written. See https://www.mongodb.com/docs/manual/administration/change-streams-production-recommendations/#sharded-clusters